dvc/repo/diff.py

```diff
@@ -23,12 +24,32 @@ def diff(self, a_rev="HEAD", b_rev=None):
 
     b_rev = b_rev if b_rev else "workspace"
     results = {}
+    missing_targets = {}
     for rev in self.brancher(revs=[a_rev, b_rev]):
         if rev == "workspace" and rev != b_rev:
             # brancher always returns workspace, but we only need to compute
             # workspace paths/checksums if b_rev was None
             continue
-        results[rev] = _paths_checksums(self)
+
+        targets_path_infos = None
+        if targets is not None:
+            # convert targets to path_infos, and capture any missing targets
+            targets_path_infos, missing_targets[rev] = _targets_to_path_infos(
+                self, targets
+            )
+
+        results[rev] = _paths_checksums(self, targets_path_infos)
+
+    if targets is not None:
+        # check for overlapping missing targets between a_rev and b_rev
+        missing_targets = list(
+            set(missing_targets[a_rev]) & set(missing_targets[b_rev])
+        )
+
+        if len(missing_targets):
+            # invalid targets supplied, raise error
+            for missing_target in missing_targets:
+                raise PathMissingError(missing_target, self)
```

Comment:
This can be simplified

```suggestion
        for target in set(missing_targets[a_rev]) & set(missing_targets[b_rev]):
            raise PathMissingError(target, self)
```

Reply:
Incorporated in https://github.com/iterative/dvc/pull/4938/commits/395167309234a0fb80cd66a416d3b52cc43f2e1e.